dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -715,8 +715,10 @@ private boolean createWorkflowsAndVersionsFromDockstoreYml(List<? extends Workfl
                     transactionHelper.continueSession().transaction(() -> {
                         Workflow workflow = result.getLeft();
                         WorkflowVersion version = result.getRight();
-                        // Automatically register a DOI
-                        automaticallyRegisterDockstoreDOI(workflow, version, user, this);
+                        if (!Boolean.TRUE.equals(wf.getDisableDoiGeneration())) {
```

Comment:
Should you also add this check to the publish workflow end point? https://github.com/dockstore/dockstore/blob/cec35ec73690bc7af06daed764d4c86687246591/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L781-L783

I suppose the check isn't needed in the refresh endpoint because manual workflows can't have their disable DOI generation field set?

https://github.com/dockstore/dockstore/blob/cec35ec73690bc7af06daed764d4c86687246591/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L310

Reply:
> Should you also add this check to the publish workflow end point?

Hmm, very good point. And it also ties into your comment about legacy workflows -- my current implementation only has access to `disableDoiGeneration` via the .dockstore.yml when a push event comes in.

To implement this correctly, I think we have to add a property to the workflow and/or workflow version. This would handle both .dockstore.yml workflows being published later, and provide users a way to opt out of legacy workflows.

@svonworl @kathy-t Thoughts?

Reply:
I'm less concerned about providing users a way to opt out for legacy workflows and arguably we could remove the feature for legacy workflows to encourage users to migrate to .dockstore.yml workflows. I think handling the case for .dockstore.yml workflows being published later is important though. I guess they could always publish it in the .dockstore.yml, but there may be users that want to do it in the UI.

An alternative is to read the .dockstore.yml file again when it's time to publish?

Reply:
> To implement this correctly, I think we have to add a property to the workflow and/or workflow version. This would handle both .dockstore.yml workflows being published later, and provide users a way to opt out of legacy workflows.

FWIW, I had not even realized that automatic DOIs support legacy workflows. 
Would also suggest considering reserving automatic DOIs for .dockstore.yml workflows to encourage a migration.

Reply:
> An alternative is to read the .dockstore.yml file again when it's time to publish (if we want to keep the feature exclusive to .dockstore.yml workflows)?

We'll know the "out-out" setting from the `.dockstore.yml`, so we probably don't need to re-read it.  But I feel like maybe I'm missing your point...

Reply:
> > An alternative is to read the .dockstore.yml file again when it's time to publish (if we want to keep the feature exclusive to .dockstore.yml workflows)?
> 
> We'll know the "out-out" setting from the `.dockstore.yml`, so we probably don't need to re-read it. But I feel like maybe I'm missing your point...

Oh, do we not store the "opt-out" setting in the version?

Reply:
> > Oh, do we not store the "opt-out" setting in the Version/VersionMetadata?
> 
> I don't think so and if we don't want to store it, we could re-read the file, but maybe that's more work than just storing it 🙂

Yeah, would prefer we store it versus trying to re-read `.dockstore.yml` later, the latter being messy and coupling us even more strongly to the notion that `.dockstore.yml` is a file in a repo somewhere (which is true now, for our "new" registration scheme, but someday maybe we'll have a different way to create entries).

Reply:
> FWIW, I had not even realized that automatic DOIs support legacy workflows. Would also suggest considering reserving automatic DOIs for .dockstore.yml workflows to encourage a migration.

I was also thinking about the PI request to retroactively issue DOIs for all workflows in the system; this could give users a chance to opt-out. I know that would be for 1.17, but this could be a step in preparation for it.

Reply:
I agree should store it outside of the .dockstore.yml, but perhaps just store it on the workflow, and not at the version level (and yes, I did originally list storing it at the version level as an option). Do we really need to offer that fine-grained control of controlling it on a per version basis? I think not, at least not for a first release.

On the other hand, the `disableDoiGeneration` in the .dockstore.yml becomes unclear; what happens if the setting is present on one tag but not another -- what do you set the workflow property to? And the current implementation doesn't let you set it to true...

Proposal 1:

- Remove the setting from .dockstore.yml
- Add a workflow-level property settable via the API to disable DOI generation. Only in the API for 1.16; UI can come later.

Proposal 2:
- Keep the property in the dockstore.yml, allow it to have both `true` and `false` values.
- Most recently processed setting wins
- Add a workflow-level property settable via the API to disable DOI generation. Only in the API for 1.16; UI can come later. The .dockstore.yml value sets this property.

Proposal 3:
- Keep the current .dockstore.yml property
- Store in the version metadata
- Also allow API to set version metadata for legacy workflows, but only optional for 1.16, as we will only generate DOIs for .dockstore.yml workflows in 1.16.

Reply:
I guess I am somewhat partial to (1) as I'm ok with the supportof  the opt-out to require contacting us to change an admin-only flag/not having a UI for it.

Although the way I had originally thought this would work is that there would be a workflow level property that was manually turned on by us (via the API) that would also initiate DOI generation. We would run this against workflows to phase in testing and retroactively initiate DOI creation. In general I prefer an approach that favors minimal API changes and then programming the API. That being said, this code is mostly written and I understand that there isn't a lot of reason to support legacy DOI generation. It's a nice-to-have to have the opt-out in the .dockstore.yml but not required.

Reply:
> I guess I am somewhat partial to (1) as I'm ok with the supportof the opt-out to require contacting us to change an admin-only flag/not having a UI for it.
> 
> Although the way I had originally thought this would work is that there would be a workflow level property that was manually turned on by us (via the API) that would also initiate DOI generation. We would run this against workflows to phase in testing and retroactively initiate DOI creation. In general I prefer an approach that favors minimal API changes and then programming the API. That being said, this code is mostly written and I understand that there isn't a lot of reason to support legacy DOI generation. It's a nice-to-have to have the opt-out in the .dockstore.yml but not required.

The allowlist of GitHub orgs was supposed to be simple. Add an org to the list and you're done. But then we added checks for not accidentally enabling orgs and the ability to opt-out, and it got more complicated.

I had thought about doing this at the API-level when starting, but it seemed like a bit of work if we were going to do it at the GitHub organization level; I probably would have needed a new table to track the orgs, the corresponding JPA entities.

The API can be admin/curator only to start with, then we can open it up to everybody in a future release.